src/robots/control/pure_pursuit_controller.cc

```diff
@@ -1,179 +1,186 @@
+{
     using namespace units::math;
 
-    if (m_wayPoints.size() > 1) {
-        bool didGetIntersection = false;
-        for (unsigned int i = 0; i < m_wayPoints.size()-1; i++) {
-
-            // path points
-            const Vector2<millimeter_t> segmentStart = m_wayPoints.at(i);
-            const Vector2<millimeter_t> segmentEnd = m_wayPoints.at(i+1);
-
-            const auto p1x = segmentStart.x() - x;
-            const auto p1y = segmentStart.y() - y;
-            const auto p2x = segmentEnd.x() - x;
-            const auto p2y = segmentEnd.y() - y;
-
-            // intersection point with circle
-            const auto dx = p2x - p1x;
-            const auto dy = p2y - p1y;
-            const auto d = sqrt(dx * dx + dy * dy);
-            const auto D = p1x * p2y - p2x * p1y;
-
-            // if the discriminant is zero -> no intersection
-            const auto discriminant = r * r * d * d - D * D;
-            if (discriminant.value() < 0) continue;
-
-            // x components of the intersection point
-            const auto x1 = (D * dy + sgn(dy.value()) * dx * sqrt(discriminant)) / (d*d);
-            const auto x2 = (D * dy - sgn(dy.value()) * dx * sqrt(discriminant)) / (d*d);
-            // y components of the intersection point
-            const auto y1 = (-D * dx + fabs(dy) * sqrt(discriminant)) / (d*d);
-            const auto y2 = (-D * dx - fabs(dy) * sqrt(discriminant)) / (d*d);
-
-            // if there is 2 intersections possible with the circle, we select the second one,
-            // as that will be closer to the end point of the segment
-            const bool validIntersection1 = (min(p1x, p2x) < x1 && x1 < max(p1x, p2x)) ||
-                                        (min(p1y, p2y) < y1 && y1 < max(p1y,p2y));
-            const bool validIntersection2 = (min(p1x, p2x) < x2 && x2 < max(p1x, p2x)) ||
-                                        (min(p1y, p2y) < y2 && y2 < max(p1y,p2y));
-
-            // we always want the latest path segment point so if we have a
-            // valid point, we delete the previous point
-            if (validIntersection1) {
-                lookaheadPoint.x() = x1 + x;
-                lookaheadPoint.y() = y1 + y;
-                didGetIntersection = true;
-            }
-
-            // if there is a valid 2. intersection point, we keep that and remove
-            // the first one
-            if (validIntersection2) {
-                if (lookaheadPoint.isnan() || fabs(x1 - p2x) > fabs(x2 - p2x) ||
-                    fabs(y1 - p2y) > fabs(y2 - p2y)) {
-                    lookaheadPoint.x() = x2 + x;
-                    lookaheadPoint.y() = y2 + y;
-                    didGetIntersection = true;
-                }
-            }
+    BOB_ASSERT(m_wayPoints.size() > 1);
+
+    std::experimental::optional<Vector2<millimeter_t>> lookAheadPoint;
+    for (unsigned int i = 0; i < m_wayPoints.size() - 1; i++) {
+        // path points
```

Comment:
It would be really nice to have a little bit more explanation of this maths

Reply:
There is *some* description in the body of the loop, but tbh I just tweaked @tenxlenx's code here so I didn't write it from scratch... @tenxlenx do you have a reference to the paper you used for this that we could stick in a comment?